docs/design/document-type.md

```diff
@@ -14,346 +14,164 @@ such as `struct`, `union`, etc. which are handled by the [Kotlin Smithy SDK](kot
-    /**
-     * Returns [SmithyNumber] associated with given [key] or `null` if element is not present or has a different type
-     */
-    fun getNumber(key: String): SmithyNumber? = getValue(key) as? SmithyNumber
-
-    /**
-     * Returns [SmithyBool] associated with given [key] or `null` if element is not present or has a different type
-     */
-    fun getBoolean(key: String): SmithyBool? = getValue(key) as? SmithyBool
-
-    /**
-     * Returns [SmithyString] associated with given [key] or `null` if element is not present or has a different type
-     */
-    fun getString(key: String): SmithyString? = getValue(key) as? SmithyString
-
-    /**
-     * Returns [SmithyArray] associated with given [key] or `null` if element is not present or has a different type
-     */
-    fun getArray(key: String): SmithyArray? = getValue(key) as? SmithyArray
-
-    /**
-     * Returns [SmithyMap] associated with given [key] or `null` if element is not present or has a different type
-     */
-    fun getMap(key: String): SmithyMap? = getValue(key) as? SmithyMap
-
-    override fun toString(): String {
-        return content.entries.joinToString(
-            separator = ",",
-            prefix = "{",
-            postfix = "}",
-            transform = {(k, v) -> """"$k":$v"""}
-        )
-    }
-}
-
-fun Boolean.toDocument() = SmithyBool(this)
-fun Number.toDocument() = SmithyNumber(this)
-fun String.toDocument() = SmithyString(this)
-
-/**
- * DSL builder for a [SmithyArray]
- */
-class DocumentArrayBuilder internal constructor() {
-    internal val content: MutableList<Document> = mutableListOf()
-
-    /**
-     * Adds [this] value to the current [SmithyArray] as [SmithyString]
-     */
-    operator fun String.unaryPlus() {
-        content.add(SmithyString(this))
+    companion object {
+        /**
+         * Overloads to create a document from possible values.
+         */
+        operator fun invoke(init: Builder.() -> Unit): Document
+        operator fun invoke(value: kotlin.Number): Document
+        operator fun invoke(value: kotlin.String): Document
+        operator fun invoke(value: kotlin.Boolean): Document
+        operator fun invoke(value: Transformable): Document
+
+        /**
+         * Unwraps a Document from a serialized value.
+         */
+        fun fromString(value: kotlin.String): Document
+        
+        /**
+         * Creates a Document list from the provided values.
+         */
+        fun listOf(vararg values: Any?): Document
     }
 
     /**
-     * Adds [this] value to the current [SmithyArray] as [SmithyBool]
+     * Inheritors that wrap all the possible values a Document can hold.
      */
-    operator fun Boolean.unaryPlus() {
-        content.add(SmithyBool(this))
-    }
-
-    /**
-     * Adds [this] value to the current [SmithyArray] as [Document]
-     */
-    operator fun Document.unaryPlus() {
-        content.add(this)
-    }
+    data class Number(val value: kotlin.Number) : Document()
```

Comment:
**Question**: Is accepting `Number` unsafe? It's not a sealed interface so 3P classes can implement it. Moreover, if/when Kotlin adds 1P `BigInteger` and `BigDecimal` implementations, they will likely inherit `Number` just like [the Java equivalents do](https://docs.oracle.com/javase/8/docs/api/java/math/BigInteger.html).

Reply:
It's not a sealed class but it's also [not an open class](https://github.com/JetBrains/kotlin/blob/6a670dc5f38fc73eb01d754d8f7c158ae0176ceb/core/builtins/native/kotlin/Number.kt#L22)

Reply:
But to Ian's point a caller can definitely extend it and shove in some bogus value, as long as they implement all the `toWhateverNumericForm()` methods.

We could instead make the constructor private, and override `invoke` etc. to take the specific subclasses.

Reply:
What am I missing here? If `Number` can't be extended from I am not seeing the concern.

Reply:
It can be though, that's what I was saying--

```kotlin
class Number2: Number() {
    override fun toByte(): Byte {
        TODO("Not yet implemented")
    }

    override fun toChar(): Char {
        TODO("Not yet implemented")
    }

    override fun toDouble(): Double {
        TODO("Not yet implemented")
    }

    override fun toFloat(): Float {
        TODO("Not yet implemented")
    }

    override fun toInt(): Int {
        TODO("Not yet implemented")
    }

    override fun toLong(): Long {
        TODO("Not yet implemented")
    }

    override fun toShort(): Short {
        TODO("Not yet implemented")
    }
}

val invalidDoc = Document.Number(Number2()) // ???
```

Reply:
oh duh, abstract classes are open. I think I need to take a break for the day.

Reply:
I definitely see Ian's point. However, I think this is overall a minor concern. Worse case we would fail serialization because we can't convert it into a builtin type.